File: modules/resource/manager/transit-network/src/nsi2interface.py
# Copyright 2014-2015 National Institute of Advanced Industrial Science and Technology
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
from datetime import datetime, timedelta
import time
import types
import log
logger = log.getLogger('tnrm:nsi2interface')
logger.info("start nsi2interface from TNRM to NSI.")

sys.path.append('/usr/share/java/')
#sys.path.append('../../common/build/jar/nsi2_common.jar')
sys.path.append('../../clientapi/build/jar/nsi2_client.jar')
sys.path.append('../libs/commons-logging-1.1.1.jar')
sys.path.append('../libs/log4j-1.2.13.jar')

sys.path.append('/opt/apache-cxf/lib/antlr-2.7.7.jar')
sys.path.append('/opt/apache-cxf/lib/aopalliance-1.0.jar')
sys.path.append('/opt/apache-cxf/lib/asm-3.3.jar')
sys.path.append('/opt/apache-cxf/lib/commons-collections-3.2.1.jar')
sys.path.append('/opt/apache-cxf/lib/commons-lang-2.6.jar')
sys.path.append('/opt/apache-cxf/lib/commons-logging-1.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-2.4.2.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-manifest.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-xjc-boolean-2.4.0.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-xjc-bug671-2.4.0.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-xjc-dv-2.4.0.jar')
sys.path.append('/opt/apache-cxf/lib/cxf-xjc-ts-2.4.0.jar')
sys.path.append('/opt/apache-cxf/lib/FastInfoset-1.2.9.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-activation_1.1_spec-1.1.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-annotation_1.0_spec-1.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-javamail_1.4_spec-1.7.1.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-jaxws_2.2_spec-1.0.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-jms_1.1_spec-1.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-servlet_3.0_spec-1.0.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-stax-api_1.0_spec-1.0.1.jar')
sys.path.append('/opt/apache-cxf/lib/geronimo-ws-metadata_2.0_spec-1.1.3.jar')
sys.path.append('/opt/apache-cxf/lib/isorelax-20030108.jar')
sys.path.append('/opt/apache-cxf/lib/jaxb-api-2.2.1.jar')
sys.path.append('/opt/apache-cxf/lib/jaxb-impl-2.2.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/jaxb-xjc-2.2.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/jettison-1.3.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-continuation-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-http-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-io-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-security-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-server-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/jetty-util-7.4.5.v20110725.jar')
sys.path.append('/opt/apache-cxf/lib/joda-time-1.6.2.jar')
sys.path.append('/opt/apache-cxf/lib/jra-1.0-alpha-4.jar')
sys.path.append('/opt/apache-cxf/lib/js-1.7R2.jar')
sys.path.append('/opt/apache-cxf/lib/jsr311-api-1.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/msv-core-2010.2.jar')
sys.path.append('/opt/apache-cxf/lib/neethi-3.0.1.jar')
sys.path.append('/opt/apache-cxf/lib/opensaml-2.4.1.jar')
sys.path.append('/opt/apache-cxf/lib/openws-1.4.1.jar')
sys.path.append('/opt/apache-cxf/lib/relaxngDatatype-20020414.jar')
sys.path.append('/opt/apache-cxf/lib/saaj-api-1.3.jar')
sys.path.append('/opt/apache-cxf/lib/saaj-impl-1.3.2.jar')
sys.path.append('/opt/apache-cxf/lib/serializer-2.7.1.jar')
sys.path.append('/opt/apache-cxf/lib/slf4j-api-1.6.1.jar')
sys.path.append('/opt/apache-cxf/lib/slf4j-jdk14-1.6.1.jar')
sys.path.append('/opt/apache-cxf/lib/spring-aop-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-asm-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-beans-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-context-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-core-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-expression-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-jms-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-tx-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/spring-web-3.0.5.RELEASE.jar')
sys.path.append('/opt/apache-cxf/lib/stax2-api-3.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/velocity-1.7.jar')
sys.path.append('/opt/apache-cxf/lib/woodstox-core-asl-4.1.1.jar')
sys.path.append('/opt/apache-cxf/lib/wsdl4j-1.6.2.jar')
sys.path.append('/opt/apache-cxf/lib/wss4j-1.6.2.jar')
sys.path.append('/opt/apache-cxf/lib/xalan-2.7.1.jar')
sys.path.append('/opt/apache-cxf/lib/xmlbeans-2.4.0.jar')
sys.path.append('/opt/apache-cxf/lib/xml-resolver-1.2.jar')
sys.path.append('/opt/apache-cxf/lib/xmlschema-core-2.0.jar')
sys.path.append('/opt/apache-cxf/lib/xmlsec-1.4.5.jar')
sys.path.append('/opt/apache-cxf/lib/xmltooling-1.3.1.jar')
sys.path.append('/opt/apache-cxf/lib/xsdlib-2010.1.jar')

#from nsi2 import *
import NSI2Interface

from java.lang import String
from java.util import ArrayList
import jarray

from reservation import Reservation

pNSA = 'urn:ogf:network:aist.go.jp:2013:nsa'
#pURI = 'https://163.220.30.173:22311/aist_upa/services/ConnectionProvider'
#pURI='http://127.0.0.1:28080/provider/services/ConnectionProvider'
#
# dummy NRM
# pURI='https://163.220.30.147:28443/provider/services/ConnectionProvider'
# dummy AG
# pURI='https://163.220.30.174:28443/provider/services/ConnectionProvider'
# real AG
pURI='https://163.220.30.174:28443/nsi2/services/ConnectionProvider'
# rNSA = 'urn:ogf:network:aist.go.jp:2013:nsa:felix'
rNSA = 'urn:ogf:network:aist.go.jp:2013:nsa'
rURI = 'https://163.220.30.145:29081/nsi2_requester/services/ConnectionRequester'
#rURI = 'https://127.0.0.1:29081/nsi2_requester/services/ConnectionRequester'
user = ''
password = ''

class NSI:
      def __init__(self):
            self.nsi = NSI2Interface(pNSA, pURI, rNSA, rURI, user, password)
            self.index = 1

      def reserve_sec(self, gid, src, dst, srcvlan, dstvlan, capacity, eros, start_sec, end_sec):
            rid = self.nsi.reserveCommit(gid, src, dst, srcvlan, dstvlan, capacity, start_sec, end_sec, eros)
            return rid

      def reserve(self, gid, sstp, dstp, svlantag, dvlantag, bw, 
                  ep_start, ep_end, eroEP):

            rid = self.nsi.reserveCommit(gid, sstp, dstp, svlantag, dvlantag,                                          bw, ep_start, ep_end, eroEP)
            self.index = self.index + 1
            logger.info("rid=%s" % (rid))
            return rid

      def modify_sec(self, gid, rid, end_sec):
            self.nsi.modifyCommit(gid, rid, end_sec)

      def modify(self, gid, rid, ep_end):
            mid = self.nsi.modifyCommit(gid, rid, ep_end)
            logger.info("mid=%s" % (mid))
            return mid

      def commit(self, rid):
            self.nsi.commit(rid)

      def abort(self, rid):
            self.nsi.abort(rid)

      def terminate(self, rid):
            self.nsi.terminate(rid)

      def provision(self, rid):
            self.nsi.provision(rid)

      def release(self, rid):
            self.nsi.release(rid)
            

if __name__ == "__main__":
      s_stp = "urn:ogf:network:aist.go.jp:2013:bi-ps"
      d_stp = "urn:ogf:network:aist.go.jp:2013:bi-aist-jgn-x"
      s_vlan = 1786
      d_vlan = s_vlan
      capacity = 100
      eros = ["urn:ogf:network:xxx:2013:stp1", "urn:ogf:network:yyy:2013:stp2"]
      gid = "TEST-GLOBAL-ID"
      nsi = NSI()

      epoch = datetime.utcfromtimestamp(0)
      start = datetime.utcnow()
      td = start - epoch
      
      start_sec = td.seconds + td.days * 24 * 3600
      end_sec = start_sec + 3600

      time.sleep(10)
      reservationId = nsi.reserve_sec(gid, s_stp, d_stp, s_vlan, d_vlan, capacity, eros, start_sec, end_sec)
      print reservationId

      time.sleep(10)
      nsi.modify_sec(gid, reservationId, end_sec)


      time.sleep(10)
      nsi.provision(reservationId)
      time.sleep(10)
      nsi.release(reservationId)
      time.sleep(10)
      nsi.terminate(reservationId)
      time.sleep(10)
      sys.exit()
      

File: modules/resource/manager/transit-network/src/NSI2Interface.java
/*
 * OGF NSI Requester API, ver. 2.0.sc
 * Copyright 2003-2013 National Institute of Advanced Industrial Science and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
# Copyright 2014-2015 National Institute of Advanced Industrial Science and Technology
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
*/

import java.util.Calendar;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.Hashtable;
import java.net.URL;

import net.glambda.nsi2.impl.EventListener;
import net.glambda.nsi2.util.NSITextDump;
import net.glambda.nsi2.util.TypesBuilder;
import nsi2.reply.QueryNotificationReply;
import nsi2.reply.QueryReply;
import nsi2.reply.ReserveCommitReply;
import nsi2.reply.ReserveReply;
import nsi2.NSI2Client;
import nsi2.SampleEventListener;

import javax.xml.bind.JAXBElement;

import org.ogf.schemas.nsi._2013._12.framework.types.ServiceExceptionType;
import org.ogf.schemas.nsi._2013._12.connection._interface.ServiceException;
import org.ogf.schemas.nsi._2013._12.connection.types.QueryNotificationConfirmedType;
import org.ogf.schemas.nsi._2013._12.connection.types.QueryNotificationType;
import org.ogf.schemas.nsi._2013._12.connection.types.QueryRecursiveResultType;
import org.ogf.schemas.nsi._2013._12.connection.types.QuerySummaryConfirmedType;
import org.ogf.schemas.nsi._2013._12.connection.types.QuerySummaryResultType;
import org.ogf.schemas.nsi._2013._12.connection.types.QueryType;
import org.ogf.schemas.nsi._2013._12.connection.types.ReservationConfirmCriteriaType;
import org.ogf.schemas.nsi._2013._12.connection.types.ReservationRequestCriteriaType;
import org.ogf.schemas.nsi._2013._12.connection.types.ScheduleType;
import org.ogf.schemas.nsi._2013._12.services.point2point.P2PServiceBaseType;
import org.ogf.schemas.nsi._2013._12.services.types.StpListType;

import org.ogf.schemas.nsi._2013._12.services.types.StpListType;
import org.ogf.schemas.nsi._2013._12.services.types.OrderedStpType;
import org.ogf.schemas.nsi._2013._12.services.point2point.P2PServiceBaseType;

import org.apache.cxf.Bus;
import org.apache.cxf.bus.spring.SpringBusFactory;
import org.apache.cxf.transport.http.HTTPConduit;

public class NSI2Interface {
    static final boolean debug = true;
    static final int DEFAULT_DURATION_HOUR = 24 * 3; // 3days
    static Hashtable<String, ReservationRequestCriteriaType> criterias =
	new Hashtable<String, ReservationRequestCriteriaType>();
    static Hashtable<String, Integer> versions =
	new Hashtable<String, Integer>();
    static Hashtable<String, Long> capacities =
	new Hashtable<String, Long>();

    String providerNSA = null;
    String providerURI = null;
    String requesterNSA = null;
    String requesterURI = null;
    String httpUser = null;
    String httpPassword = null;
    // long replyWait = 60 * 1000L; // 60 sec
    long replyWait = 300 * 1000L; // 60 sec

    EventListener listener = null;
    NSI2Client client = null;
    boolean hasRequester = false;

    private final String DEFAULT_GLOBAL_RESERVATION_ID = "global";
    private final String DEFAULT_DESCRIPTION = "nsi2 test";

    public NSI2Interface
	(String providerNSA, String providerURI, 
	 String requesterNSA, String requesterURI,
	 String httpUser, String httpPassword)
	throws Exception
    {
	this.providerNSA = providerNSA;
	this.providerURI = providerURI;
	this.requesterNSA = requesterNSA;
	this.requesterURI = requesterURI;
	this.httpUser = httpUser;
	this.httpPassword = httpPassword;

	if (requesterURI != null) hasRequester = true;
        listener = new SampleEventListener();
	client = getNSI2Client();
    }

    private NSI2Client getNSI2Client() 
	throws Exception
    {
	NSI2Client c = null;
	String oauth = null;

	if (httpUser != null && httpPassword == null) oauth = httpUser;
        if (oauth == null) {
	    c = new NSI2Client
		(providerNSA, providerURI, requesterNSA, requesterURI,
		 replyWait, httpUser, httpPassword, listener);
	} else {
	    c = new NSI2Client
		(providerNSA, providerURI, requesterNSA, requesterURI,
		 replyWait, oauth, listener);
        }
	return c;
    }

    private int getNewVersion(String reservationId) 
    {
        Integer it = versions.get(reservationId);
	if (it == null) {
	    it = new Integer(0);
	    versions.put(reservationId, it);
	} else {
	    it ++;
	}
	if (debug) System.out.println("******************* value=" + it);
	return it.intValue();
    }


    public static P2PServiceBaseType getP2PServiceBaseType(List<Object> list)
    {
	Object o = list.get(0);
	if (o instanceof JAXBElement) {
	    Object ele = ((JAXBElement<?>) o).getValue();
	    if (ele instanceof P2PServiceBaseType) 
		return (P2PServiceBaseType) ele;
	}

	return null;
    }

    public String reserveCommit
	(String gid, String sSTP, String dSTP, int sVlan, int dVlan, long capacity,
	 int start, int end, String[] eros)
	throws Exception 
    {
	ReservationRequestCriteriaType criteria = makeReservationCriteria
	    (gid, sSTP, dSTP, sVlan, dVlan, capacity, start, end);

	P2PServiceBaseType p2p = getP2PServiceBaseType(criteria.getAny());
	if (p2p != null) {
	    StpListType erolist = p2p.getEro();
	    if (eros != null) {
		int size = eros.length;
		if (erolist == null && size != 0) {
		    erolist = new StpListType();
		    p2p.setEro(erolist);
		}
		List<OrderedStpType> oero = erolist.getOrderedSTP();

		for (int i = 0; i < size; i ++) {
		    if (debug) System.out.println("******************* ero[" + i + "]=" + eros[i]);
		    OrderedStpType ost = new OrderedStpType();
		    ost.setStp(eros[i]);
		    ost.setOrder(i);
		    
		    oero.add(ost);
		}
	    } else {
		if (debug) System.out.println("******************* ero os null");
	    }
	}

	try {
	    String reservationId = reserve(gid, criteria);
	    if (debug) System.out.println("******************* reservationId:" + reservationId);
	    commit(reservationId);

	    criterias.put(reservationId, criteria);
	    capacities.put(reservationId, new Long(capacity));
	    return reservationId;
	} catch (Exception e) {
	    if (debug) System.out.println("******************* ex:" + e);
	    throw e;
	}
    }

    public ReservationRequestCriteriaType makeReservationCriteria
	(String gid, String sSTP, String dSTP, int sVlan, int dVlan, long capacity,
	 int startTime, int endTime, String[] ero)
    {
	ReservationRequestCriteriaType criteria = makeReservationCriteria
	    (gid, sSTP, dSTP, sVlan, dVlan, capacity, startTime, endTime);

	if (ero == null || ero.length == 0) return criteria;

	List<Object> objs = criteria.getAny();
	if (! objs.isEmpty()) {
	    Object obj = objs.get(0);
	    if (! (obj instanceof P2PServiceBaseType)) 
		return criteria;

	    StpListType slt = ((P2PServiceBaseType) obj).getEro();
	    if (slt == null) {
		slt = new StpListType();
		((P2PServiceBaseType) obj).setEro(slt);
	    }
	    List<OrderedStpType> osp= slt.getOrderedSTP();

	    for (int i = 0; i < ero.length; i++) {
		String stp = ero[i];
		if (debug) System.out.println("order=" + i +
					      ", stp=" + stp);

		OrderedStpType ostp = new OrderedStpType();
		ostp.setOrder(i);
		ostp.setStp(stp);

		osp.add(ostp);
	    }
	}
	return criteria;
    }

    public ReservationRequestCriteriaType makeReservationCriteria
	(String gid, String sSTP, String dSTP, int sVlan, int dVlan, long capacity,
	 int startTime, int endTime)
    {
        String srcstp = sSTP;
        String deststp = dSTP;

        Calendar start = Calendar.getInstance();
	long currentTime = start.getTimeInMillis();
	long lstart = ((long) startTime) * 1000;
	long lend = ((long) endTime) * 1000;
	if (lstart >= currentTime) start.setTimeInMillis(lstart);
        Calendar end = (Calendar) start.clone();
	if (lend <= 1000L*3600*3) 
	    end.add(Calendar.HOUR, DEFAULT_DURATION_HOUR);
	else
	    end.setTimeInMillis(lend);
	
        int srcvlan = sVlan;
        int destvlan = dVlan;
        long bandwidth = capacity; // [Mbps]
	
        ScheduleType schedule = TypesBuilder.makeScheduleType(start, end);
        ReservationRequestCriteriaType criteria =
	    TypesBuilder.makeReservationRequestCriteriaType
	    (schedule, srcstp, srcvlan, deststp, destvlan, capacity);

	criteria.setVersion(0);
	return criteria;
    }

    public String reserve(String gid, ReservationRequestCriteriaType criteria) 
	throws Exception 
    {
	String reservationId = null;
        // String globalReservationId = DEFAULT_GLOBAL_RESERVATION_ID;
        String globalReservationId = gid;
        String description = DEFAULT_DESCRIPTION;
        if (debug) showMessage(criteria, NSITextDump.toString(criteria));

	ReserveReply reply = null;
	try {
	    reply = client.reserve
		(reservationId, globalReservationId, description, criteria);
	} catch (Exception e) {
	    if (debug) System.out.println("******************* ex:" + e);
	    throw e;
	}

        if (reply.getConfirm() != null) {
            ReservationConfirmCriteriaType conf = reply.getConfirm();
            if (debug) showMessage("ConnectionID", reply.getConnectionId());
            if (debug) showMessage(conf, NSITextDump.toString(conf));
	    int version = conf.getVersion();
	    versions.put(reply.getConnectionId(), new Integer(version));
        } else if (reply.getServiceException() != null) {
            if (debug) showMessage
			   (reply.getServiceException(),
			    NSITextDump.toString(reply.getServiceException()));
            if (debug) showMessage
			   (reply.getConnectionStates(),
			    NSITextDump.toString(reply.getConnectionStates()));

	    ServiceExceptionType t = reply.getServiceException();
	    List<ServiceExceptionType> l = t.getChildException();
	    String s = "ServiceException: \n";
	    for (ServiceExceptionType i: l) {
		s += "    :" + i.getText() + "\n";
	    }
	    throw new Exception (s);
	}
        return reply.getConnectionId();
    }

    public void commit(String reservationId) 
	throws Exception 
    {
        ReserveCommitReply reply = client.reserveCommit(reservationId);
        if (reply.getServiceException() == null) {
            if (debug) System.out.println("ReserveCommitConfirmed");
        } else if (reply.getServiceException() != null) {
            if (debug) showMessage
			   (reply.getServiceException(),
			    NSITextDump.toString(reply.getServiceException()));
            if (debug) showMessage
			   (reply.getConnectionStates(),
			    NSITextDump.toString(reply.getConnectionStates()));
        }
    }

    public void abort(String reservationId) 
	throws Exception 
    {
        client.reserveAbort(reservationId);
    }

    public String modify (String gid, String reservationId, int end_time_sec)
	throws Exception 
    {
        // String globalReservationId = DEFAULT_GLOBAL_RESERVATION_ID;
        String globalReservationId = gid;
        String description = DEFAULT_DESCRIPTION;

        ReservationRequestCriteriaType old = 
	    (ReservationRequestCriteriaType) criterias.get(reservationId);
	if (old == null) {
	    throw new ServiceException
		("No such reservation, id=" + reservationId);
	}
        ScheduleType oldSchedule = old.getSchedule();
        Calendar oldStart = oldSchedule.getStartTime();

        Calendar start = Calendar.getInstance();
	
	long lend = ((long) end_time_sec) * 1000;
        Calendar end = Calendar.getInstance();
	end.setTimeInMillis(lend);

        // Calendar end = (Calendar) start.clone();
        // end.add(Calendar.HOUR, DEFAULT_DURATION_HOUR);
        ScheduleType schedule = TypesBuilder.makeScheduleType(oldStart, end);

	long capacity = capacities.get(reservationId).longValue();

        ReservationRequestCriteriaType criteria =
	    TypesBuilder.makeReservationRequestCriteriaType(schedule, capacity);
        criteria.setVersion(getNewVersion(reservationId));
	if (debug) showMessage(criteria, NSITextDump.toString(criteria));

        ReserveReply reply = client.reserve
	    (reservationId, globalReservationId, description, criteria);

        if (reply.getConfirm() != null) {
            ReservationConfirmCriteriaType conf = reply.getConfirm();
            if (debug) showMessage(conf, NSITextDump.toString(conf));
	    int version = conf.getVersion();
	    versions.put(reservationId, new Integer(version));
        } else if (reply.getServiceException() != null) {
            if (debug) showMessage
			   (reply.getServiceException(),
			    NSITextDump.toString(reply.getServiceException()));
            if (debug) showMessage
			   (reply.getConnectionStates(),
			    NSITextDump.toString(reply.getConnectionStates()));
        }

	return reservationId;
    }


    public String modifyCommit (String gid, String reservationId, int end_time_sec)
	throws Exception 
    {
	String mid = modify(gid, reservationId, end_time_sec);
        commit(reservationId);
	return mid;
    }

    public void query(String reservationId) throws Exception {
        QueryType queryType = new QueryType();
        queryType.getConnectionId().add(reservationId);
        {
            // QuerySummarySync
            QuerySummaryConfirmedType summary = 
		client.querySummarySync(queryType);
            if (debug) showMessage
			   ("QuerySummary (sync)", 
			    NSITextDump.toString(summary));
        }
        if (hasRequester) {
            // QuerySummary (async)
            QueryReply reply = client.querySummary(queryType);
            if (reply.getSummary() != null) {
                List<QuerySummaryResultType> summary = reply.getSummary();
                if (debug) showMessage
			       ("QuerySummary (async)", 
				NSITextDump.toStringQuerySummayList(summary));
            } else {
                if (debug) showMessage
			       ("QuerySummary (async)", 
				NSITextDump.toString(reply.getException()));
            }
        }
        if (hasRequester) {
            // QueryRecursive (async)
            QueryReply reply = client.queryRecursive(queryType);
            if (reply.getRecursive() != null) {
                List<QueryRecursiveResultType> recursive = reply.getRecursive();
                if (debug) showMessage
			       ("QueryRecursive (async)",
				NSITextDump.toStringQueryRecursiveList(recursive));
            } else {
                if (debug) showMessage
			       ("QueryRecursive (async)", 
				NSITextDump.toString(reply.getException()));
            }
        }
    }

    public void queryNotification (String reservationId) 
	throws Exception 
    {
        QueryNotificationType type = new QueryNotificationType();
        type.setConnectionId(reservationId);
        {
            // QueryNotification (sync)
            QueryNotificationConfirmedType conf = client.queryNotificationSync(type);
            if (debug) showMessage
			   ("QueryNotification (sync)", 
			    NSITextDump.toString(conf));
        }
        if (hasRequester) {
            // QueryNotification (async)
            QueryNotificationReply reply = client.queryNotification(type);
            if (reply.getConfirmed() != null) {
                if (debug) showMessage
			       ("QueryNotification (async)", 
				NSITextDump.toString(reply.getConfirmed()));
            } else {
                if (debug) showMessage
			       ("QueryNotification (async)", 
				NSITextDump.toString(reply.getException()));
            }
        }
    }

    public void provision(String reservationId) 
	throws Exception 
    {
	client.provision(reservationId);
    }

    public void release(String reservationId) 
	throws Exception 
    {
	client.release(reservationId);
    }

    public void terminate(String reservationId) 
	throws Exception 
    {
	client.terminate(reservationId);

	criterias.remove(reservationId);
	versions.remove(reservationId);
	capacities.remove(reservationId);
    }

    static private void showMessage(String header, String txt) 
    {
        System.out.println("**** " + header + " ****");
        System.out.println(txt);
        System.out.println();
    }

    static private void showMessage(Object o, String txt) 
    {
        if (o != null) {
            showMessage(o.getClass().getSimpleName(), txt);
        } else {
            showMessage("", txt);
        }
    }

    public static void main(String[] args)
    {
	try {
	    NSI2Interface nsi = new NSI2Interface
		("urn:ogf:network:aist.go.jp:2013:nsa",
		 // "https://127.0.0.1:22311/aist_upa/services/ConnectionProvider",
		 "http://127.0.0.1:28080/provider/services/ConnectionProvider",
		 "urn:ogf:network:aist.go.jp:2013:nsa",
		 "https://127.0.0.1:29081/nsi2_requester/services/ConnectionRequester",
		 null,
		 null);

	    String[] eros = new String[2];
	    eros[0] = "rn:ogf:network:xxx:2013:ero";
	    eros[1] = "rn:ogf:network:yyy:2013:ero";
	    String gid = "global-id";
	    
	    String id = nsi.reserveCommit
		(gid, 
		 "urn:ogf:network:aist.go.jp:2013:bi-ps",
		 "urn:ogf:network:aist.go.jp:2013:bi-aist-jgn-x",
		 1783, 1783, 400, 0, 0, eros);
	    nsi.provision(id);

	    Thread.sleep(10 * 1000);
	    nsi.modifyCommit(gid, id, 0);

	    Thread.sleep(10 * 1000);
	    nsi.release(id);

	    Thread.sleep(10 * 1000);
	    nsi.provision(id);

	    Thread.sleep(10 * 1000);
	    nsi.terminate(id);
	} catch (Exception ex) {
	    ex.printStackTrace();
	}
    }
}
